src/backend/allocator/swapchain.rs

```diff
@@ -0,0 +1,164 @@
+///
+/// ## How am I supposed to use this?
+///
+/// To do proper buffer management, most compositors do so called double-buffering.
+/// Which means you use two buffers, one that is currently presented (the front buffer)
+/// and one that is currently rendered to (the back buffer). After each rendering operation
+/// you swap the buffers around, the old front buffer becomes the new back buffer, while
+/// the new front buffer is displayed to the user. This avoids showing the user rendering
+/// artifacts doing rendering.
+///
+/// There are also reasons to do triple-buffering, e.g. if you swap operation takes a
+/// unspecified amount of time. In that case you have one buffer, that is currently
+/// displayed, one that is done drawing and about to be swapped in and another one,
+/// which you can use to render currently.
+///
+/// Re-using and managing these buffers becomes increasingly complex the more buffers you
+/// introduce, which is where `Swapchain` comes into play.
+///
+/// `Swapchain` allocates buffers for you and transparently re-created them, e.g. when resizing.
+/// All you tell the swapchain is: *"Give me the next free buffer"* (by calling [`acquire`](Swapchain::acquire)).
+/// You then hold on to the returned buffer during rendering and swapping and free it once it is displayed.
+/// Efficient re-use of the buffers is done by the swapchain.
+///
+/// If you have associated resources for each buffer, that can also be re-used (e.g. framebuffer Handles for a `DrmDevice`),
+/// you can store then in the buffer slots userdata, where it gets freed, if the buffer gets allocated, but
+/// is still valid, if the buffer was just re-used. So instead of creating a framebuffer handle for each new
+/// buffer, you can skip creation, if the userdata already contains a framebuffer handle.
+pub struct Swapchain<A: Allocator<B>, B: Buffer, U: 'static> {
+    /// Allocator used by the swapchain
+    pub allocator: A,
+
+    width: u32,
+    height: u32,
+    format: Format,
+
+    slots: [Slot<B, U>; SLOT_CAP],
+}
+
+/// Slot of a swapchain containing an allocated buffer and its userdata.
+///
+/// Can be cloned and passed around freely, the buffer is marked for re-use
+/// once all copies are dropped. Holding on to this struct will block the
+/// buffer in the swapchain.
+pub struct Slot<B: Buffer, U: 'static> {
+    buffer: Arc<Option<B>>,
+    acquired: Arc<AtomicBool>,
+    userdata: Arc<Mutex<Option<U>>>,
+}
+
+impl<B: Buffer, U: 'static> Slot<B, U> {
+    /// Set userdata for this slot.
+    pub fn set_userdata(&self, data: U) -> Option<U> {
+        self.userdata.lock().unwrap().replace(data)
+    }
+
+    /// Retrieve userdata for this slot.
+    pub fn userdata(&self) -> MutexGuard<'_, Option<U>> {
+        self.userdata.lock().unwrap()
+    }
+
+    /// Clear userdata contained in this slot.
+    pub fn clear_userdata(&self) -> Option<U> {
+        self.userdata.lock().unwrap().take()
+    }
```

Comment:
Ok nitpick time, there is a lot of redundancy between these functions, mostly exposing only `userdata()` would be enough, no? The two others are trivially implemented in terms of it, to the point that maybe the user could just do it?

Reply:
Yeah, you are probably right.